### PR TITLE
[Quant][PT2.0] fix issues for rearranging weight observer for decomposed linear

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
**Summary**
Linear is decomposed to `t - addmm/mm` after `dynamo.export`. And weight's observer is inserted between `t` and `addmm/mm` in the first place. `_rearrange_weight_observer_for_addmm()` is then called to move the observer between weight and `t`.
```
    before:
         weight - t - observer \
           input - observer - addmm/mm
    after:
         weight - observer - t \
           input - observer - addmm/mm
```
We found two issues of `_rearrange_weight_observer_for_addmm()`:
- It does not call `m.recompile()` in the end, so it does not function correctly.
- It does not support `aten.mm.default` which is from decomposed linear without bias.

This PR fixes the two issues and renames the function to `_rearrange_weight_observer_for_decomposed_linear`.

**Test plan**
python test/test_quantization.py -k test_rearrange_weight_observer_for_decomposed_linear

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168. Please take a look. Thanks!

**Comment by Xia-Weiwen** [comment]
Hi @supriyar. Could you please take a look? Thanks.

**Comment by vkuzo** [comment]
Thanks for this!

I feel like a more proper way to handle this would be to register the pattern with `mm` and `transpose`, and ensure that gets matched before the `mm` pattern.  @andrewor14 , do you know an eng estimate on that?  If that would be hard, I'd vote for landing the current PR and adding a TODO.  If that's easy, I'd vote for just doing that.

thoughts?

**Comment by andrewor14** [comment]
> Thanks for this!
> 
> I feel like a more proper way to handle this would be to register the pattern with `mm` and `transpose`, and ensure that gets matched before the `mm` pattern. @andrewor14 , do you know an eng estimate on that? If that would be hard, I'd vote for landing the current PR and adding a TODO. If that's easy, I'd vote for just doing that.
> 
> thoughts?

I think we would still insert an observer before all inputs of `mm`, so we would still have to manually remove the observer between `t` and `mm`. If we need graph post processing either way, I think the current approach is simpler. I believe that's also what @jerryzh168 intended for addmm as well.

**Review by andrewor14 — APPROVED** [review]
Looks good, just a few small nits. Thanks for fixing this!

File: torch/ao/quantization/_quantize_pt2e.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def prepare_pt2e(
 
     # TODO: remove hack when we have better support for pattern matching
     # move around the observer for addmm
```
Comment (andrewor14) [inline torch/ao/quantization/_quantize_pt2e.py:42]: @vkuzo Looks like there's already a TODO for this

**Comment by vkuzo** [comment]
> > Thanks for this!
> > I feel like a more proper way to handle this would be to register the pattern with `mm` and `transpose`, and ensure that gets matched before the `mm` pattern. @andrewor14 , do you know an eng estimate on that? If that would be hard, I'd vote for landing the current PR and adding a TODO. If that's easy, I'd vote for just doing that.
> > thoughts?
> 
> I think we would still insert an observer before all inputs of `mm`, so we would still have to manually remove the observer between `t` and `mm`. If we need graph post processing either way, I think the current approach is simpler. I believe that's also what @jerryzh168 intended for addmm as well.

If we're trying to match F.linear, and the pattern includes the transpose, there should be no graph post-processing needed.  The post processing seems to be here because the matched pattern is not actually F.linear but the subset of it.

Can we add a TODO to new code added in this PR to track switching it to the more proper pattern in the future?

**Comment by andrewor14** [comment]
> If we're trying to match F.linear, and the pattern includes the transpose, there should be no graph post-processing needed. The post processing seems to be here because the matched pattern is not actually F.linear but the subset of it.

What I meant is we still need to add an observer before F.linear for the input path. If we already match `t - F.linear` for the weight path, then we can't match the same `F.linear` again separately with our current pattern matching code. I agree this is a hack, but until we have better pattern matching capability (which seems like a large effort) this hack seems the simplest for now.

File: test/quantization/fx/test_quantize_pt2e.py [hunk 2]
```diff
@@ -126,6 +128,69 @@ def forward(self, x):
...
+                    self.assertTrue(target == torch.ops.aten.t.default, error_msg)
+
+            # 2. Check m.code to ensure `m.recompile()` is called.
+            # If the arrangement works, the code should contain
+            #   addmm_default = torch.ops.aten.addmm.default(_param_constant1, activation_post_process_0, t_default_1);
+            # or
+            #   mm_default = torch.ops.aten.mm.default(activation_post_process_0, t_default_1);
+            # Otherwise, the last argument of addmm or mm is 'activation_post_process_1'.
+            expected_code = 'activation_post_process_0, t_default_1'
+            self.assertTrue(expected_code in m.code, error_msg)
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_pt2e.py:192]: I am also wondering if there is a better way to check this. I cannot think of any. Any suggestion would be much appreciated!

[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_pt2e.py]: Thanks. It's fixed. Now we compare `m.code` before and after we call `m.recompile()` explicitly. If the two versions of code match, test passes.

File: torch/ao/quantization/_pt2e/utils.py [hunk 3]
```diff
@@ -97,23 +97,29 @@ def _fuse_conv_bn_(m: GraphModule) -> None:
...
          weight - t - observer \
-          input - observer - addmm
+         input - observer - addmm/mm
     after:
          weight - observer - t \
-           input - observer - addmm
+         input - observer - addmm/mm
     """
+    aten = torch.ops.aten
+    op_to_weight_index = {
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply torch/ao/quantization/_pt2e/utils.py]: Thanks. It's fixed.

File: torch/ao/quantization/_pt2e/utils.py [hunk 4]
```diff
@@ -97,23 +97,29 @@ def _fuse_conv_bn_(m: GraphModule) -> None:
...
     """
+    Linear is decomposed to `t - addmm` (w/ bias) or `t - mm` (w/o bias)
     before:
          weight - t - observer \
-          input - observer - addmm
+         input - observer - addmm/mm
     after:
          weight - observer - t \
-           input - observer - addmm
+         input - observer - addmm/mm
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply torch/ao/quantization/_pt2e/utils.py]: Thanks. It's fixed.

**Comment by Xia-Weiwen** [comment]
Hi @andrewor14. I have made changes per your comments. Could you please take a look again? Thanks.

**Comment by andrewor14** [comment]
@Xia-Weiwen Thanks, looks great! Can you also add a TODO per @vkuzo's comment, to match the pattern `t - addmm/mm` in the future once we have better pattern matching capability, so we don't have to do this hack anymore?

**Comment by Xia-Weiwen** [comment]
> @Xia-Weiwen Thanks, looks great! Can you also add a TODO per @vkuzo's comment, to match the pattern `t - addmm/mm` in the future once we have better pattern matching capability, so we don't have to do this hack anymore?

OK. It's added.

**Pull request closed without merging**